### PR TITLE
Skip filter state update when composite filter is in upstream

### PR DESCRIPTION
**Pull request by yanjunxiang-google**
This is to address issue: https://github.com/envoyproxy/envoy/issues/33343

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/filters/http/composite/filter.h [hunk 1]
```diff
@@ -109,6 +110,8 @@ class Filter : public Http::StreamFilter,
     }
   }
 
+  bool isUpstream() { return is_upstream_; }
```
Comment (juanmolle) [inline source/extensions/filters/http/composite/filter.h:113]: this method can be const.

**Review by juanmolle — COMMENTED** [review]
Looks Good, probably adding a line indicating in the documentation this is not emitted when the filter is config as upstream could help. Just in case someone wait for the info that will never be publish https://github.com/envoyproxy/envoy/blob/main/docs/root/configuration/http/http_filters/composite_filter.rst?plain=1#L21

File: docs/root/configuration/http/http_filters/composite_filter.rst [hunk 2]
```diff
@@ -20,6 +20,7 @@ incremented.
 
 This filter adds a map of the delegated filter name (of the action that is matched )and the root config filter name to the filter state with key
 ``envoy.extensions.filters.http.composite.matched_actions``
+This filter state is not emitted when the filter is configed in the upstream filter chain.
```
Comment (juanmolle) [inline docs/root/configuration/http/http_filters/composite_filter.rst:23]: configed is ok or should be configured?

Comment (juanmolle) [reply docs/root/configuration/http/http_filters/composite_filter.rst:23]: never minded, configed is past participle of config, sorry I'm not english native

**Review by juanmolle — APPROVED** [review]
LGTM

**Comment by juanmolle** [comment]
envoy 1.30 was released. does this PR got sens now? if it does and get merged in 1.31, then eventually will break compatibility right?

**Comment by yanjunxiang-google** [comment]
> envoy 1.30 was released. does this PR got sens now? if it does and get merged in 1.31, then eventually will break compatibility right?

composite filter status_upstream is alpha : https://github.com/envoyproxy/envoy/blob/313b6fb7cf0f806e74a2d42c93e7c1fcccce2391/source/extensions/extensions_metadata.yaml#L284

This gives us flexibility to make some upstream specific changes.

**Comment by htuch** [comment]
Doesn't the issue suggest we could support upstream filter state? Or is this "we'll implement it when needed"?

**Comment by yanjunxiang-google** [comment]
> Doesn't the issue suggest we could support upstream filter state? Or is this "we'll implement it when needed"?

This PR implements the latter.

**Pull request merged**